barretenberg/cpp/src/barretenberg/ecc/fields/field_docs.md

```diff
@@ -0,0 +1,190 @@
+2. Implementation using Intel ADX. It allows simultaneous use of two addition-with carry operations (adox and adcx) on two separate CPU gates (units of execution that can work simultaneously on the same core), which almost halves the time spent adding up the results of uint64_t multiplication.
+
+Implementation for WASM:
+
+We use 9 29-bit limbs for computation (storage stays the same) and we change the Montgomery form. The reason for a different architecture is that WASM doesn't have:
+1. 128-bit result 64*64 bit multiplication
+2. 64-bit addition with carry
+
+In the past we implemented a version with 32-bit limbs, but as a result, when we accumulated limb products we always had to split 64-bit results of 32-bit multiplication back into 32-bit chunks. Had we not, the addition of 2 64-bit products would have lost the carry flag and the result would be incorrect. There were 2 issues with this:
+1. This spawned in a lot of masking operations
+2. We didn't use more efficient algorithms for squaring, because multiplication by 2 of intermediate products would once again overflow.
+
+Switching to 9 29-bit limbs increased the number of multiplications from 136 to 171. However, since the product of 2 limbs is 58 bits, we can safely accumulate 64 of those before we have to reduce. This allowed us to get rid of a lot of intermediate masking operations, shifts and additions, so the resulting computation turned out to be more efficient. 
+
+## Interaction of field object with other objects
+Most of the time field is used with uint64_t or uint256_t in our codebase, but there is general logic of how we generate field elements from integers:
+1. Converting from signed int takes the sign into account. It takes the absolute value, converts it to montgomery and then negates the result if the original value was negative
+2. Unsigned integers ( <= 64 bits) are just converted to montgomery
+3. uint256_t and uint512_t: 
+    1. Truncate to 256 bits
+    2. Subtract the modulus until the value is within field
+    3. Convert to montgomery
+
+Conversion from field elements exists only to unsigned integers and bools. The value is converted from montgomery and appropriate number of lowest bits is used to initialize the value. 
+
+**N.B.** Functions for converting from uint256_t and back are not bijective, since values \f$ \ge p\f$ will be reduced.
+
+## Field parameters
+
+The field template is instantiated with field parameter classes, for example, class bb::Bn254FqParams. Each such class contains at least the modulus (in 64-bit and 29-bit form), r_inv and 2 versions of r_squared (64-bit and WASM/29-bit version). r_squared and other parameters (such as cube_root, primitive_root and coset_generators) are defined for wasm separately, becuase the values represent an element already in Montgomery form.
```

Comment:
Can you remind the reader what `r_inv` is, why we need `r_square` also why we don't also store `R`, `R_square` since it's also constant right?

Reply:
We store R_square. We don't store R